File: bms/go/go-edlib/SorensenDiceCoefficient/SorensenDiceCoefficient.go
package test 

import (
)

var (
)

const (
)

func SorensenDiceCoefficient(str1, str2 string, splitLength int) float32 {
	if str1 == "" && str2 == "" {
		return 0
	}
	shingle1 := Shingle(str1, splitLength)
	shingle2 := Shingle(str2, splitLength)

	intersection := float32(0)
	for i := range shingle1 {
		if _, ok := shingle2[i]; ok {
			intersection++
		}
	}
	return 2.0 * intersection / float32(len(shingle1)+len(shingle2))
}

func Shingle(s string, k int) map[string]int {
	m := make(map[string]int)
	if s != "" && k != 0 {
		runeS := []rune(s)

		for i := 0; i < len(runeS)-k+1; i++ {
			m[string(runeS[i:i+k])]++
		}
	}
	return m
}

File: translations_only/fail/go/go-edlib/SorensenDiceCoefficient/gemini/Hinted/SorensenDiceCoefficient.rs

use std::collections::HashMap;

fn sorensen_dice_coefficient(str1: &str, str2: &str, split_length: usize) -> f32 {
    if str1.is_empty() && str2.is_empty() {
        return 0.0;
    }
    let shingle1 = shingle(str1, split_length);
    let shingle2 = shingle(str2, split_length);

    let mut intersection = 0.0;
    for key in shingle1.keys() {
        if shingle2.contains_key(key) {
            intersection += 1.0;
        }
    }
    2.0 * intersection / (shingle1.len() as f32 + shingle2.len() as f32)
}

fn shingle(s: &str, k: usize) -> HashMap<String, isize> {
    let mut m = HashMap::new();
    if !s.is_empty() && k != 0 {
        let rune_s = s.chars().collect::<Vec<char>>();

        for i in 0..rune_s.len() - k + 1 {
            let key = rune_s[i..i + k].iter().collect::<String>();
            *m.entry(key).or_insert(0) += 1;
        }
    }
    m
}
